### ISSUE TITLE
Remove logging framework

### ISSUE THREAD

**Issue by vwkd** [denoland/std#6124]
I'd suggest to remove the log module.

It isn't cross platform compatible, has the lowest JSR score of all `@std` packages with 88%, and is still unstable. The API has not evolved since originally porting it from Go, the API doesn't really fit JavaScript with magical global state that's influenced by the `setup` function, which the user must take care to always call first before any implementing library.

It's the only framework in the standard library. There are different ways to do logging and not one clear solution. There are various third-party logging frameworks already, often better maintained and cross platform. Similar arguments why a HTTP middleware framework hasn't been added yet #1295.

**Comment by iuioiua** [comment]
I agree. I very much dislike `@std/log`. Discussions with Yoshiya (when I maintained this project) aligned with replacing the entire package with something better.

**Comment by halvardssm** [comment]
I partially agree, I don’t mind the logger changing, but I am finding it very useful to have a standard logger library that can log to the console, a file and whatever pluggable custom logger I would find or make.

**Comment by BlackAsLight** [comment]
I am in favour of it being changed instead of removed. I'm not 100% familiar with how it works, but things I'd like to see in a logging framework is:
- The ability for libraries to implement it in their code that doesn't add additional computational strait if the main module isn't using it.
  - This could maybe be archived though accepting simple types in `log.debug()` or any type that implements a custom symbol like `Symbol.Log`, where calling the symbol on the object will have it handle how to stringify the object.
- For the main module to decide completely how the output of the logger is displayed instead of opting for JSON.
- A cron like expression that triggers to allow rotation of where the logs are going. For example if it's going to a file, lets one switch to a new file after x hours.
- Many of the settings for the main module be configured via environment variables. This can remove people's disdain for the need to call `log.setup()` Like:
  - Default level being logged.
  - Library specific level being logged.
  - Location the logs are logged: "stdout", "stderr", a directory/file name, "null", or let runtime code decide.

**Comment by chyzwar** [comment]
What is the best alternative? Last time I tried to use pino it was not working.
As user, I would want to know what is recommended logging framework in deno.

**Comment by kt3k** [comment]
LogTape seems relatively popular to me https://github.com/dahlia/logtape
It officially supports Deno, and is available from JSR https://jsr.io/@logtape/logtape

**Comment by chyzwar** [comment]
Logtape is not even a year old and only have one maintainer. It only has 510 stars. For me, this is the definition of niche and very early stage. Is there is a way see download stats in JSR?

**Comment by chyzwar** [comment]
> I think if the quality of the design is good enough, the package is good enough to be adopted anyway.

I disagree. The maintainer could disappear tomorrow, and you will end up with hundreds of call sites to a library that is no longer maintained. Logtape support both `structured logging` and `lazy evaluation`  these features are in my opinion are bad ideas and need to be carefully evaluated if you do not want to have another Log4Shell moment. 

For me the biggest benefit of Deno is DX.  When you use Deno you no longer need to pick random quality libraries. You have a great standard library that can cover common use cases. Things like logger are fundamental to developing of any application. I suspect that logger is one of the more used packages from std.

**Comment by kt3k** [comment]
> Last time I tried to use pino it was not working.

`pino` now seems working to me in Deno (The below worked):

```ts
import pino from "npm:pino";
const logger = pino();
logger.info("hello world");
const child = logger.child({ a: "property" });
child.info("hello child!");
```

`pino` seems great in its support policy https://github.com/pinojs/pino/blob/main/docs/lts.md

**Comment by kt3k** [comment]
Note: We are making OpenTelemery builtin to Deno runtime https://docs.deno.com/runtime/fundamentals/open_telemetry/

**Comment by kt3k** [comment]
Side note: pino's file transport compatibility was fixed in https://github.com/denoland/deno/pull/27378

**Comment by alexgleason** [comment]
I created a custom JSON "logger" framework after reading this thread, much in the spirit of Vanilla JS: https://jsr.io/@soapbox/logi 🤣

**Comment by BlackAsLight** [comment]
I'd also like the continuation of `@std/log`. I figured out a cool way to have ignored logs have very little overhead with short circuiting. 

```ts
import { delay } from "@std/async";
import { info } from "@std/log";

info?.("Hello World");

info?.(async function () {
  await delay(1000);
  return "Bye World";
}());
```

Essentially at startup info will either be a function that accepts a `string` or `Promise<string>` or be undefined and just skipped over very fast.

**Comment by funnylookinhat** [comment]
Here's something I created as a thought experiment from this thread.

https://jsr.io/@funnylookinhat/logosaurus@1.0.0

**Comment by dahlia** [comment]
Hi everyone, I'm the author of [LogTape](https://logtape.org/). I noticed some concerns about LogTape in this discussion that I'd like to address.

### Regarding security concerns

> Logtape support both `structured logging` and `lazy evaluation` these features are in my opinion are bad ideas and need to be carefully evaluated if you do not want to have another Log4Shell moment.

This comparison isn't technically accurate. The Log4Shell vulnerability was possible because Log4j would interpret and execute strings in log messages through JNDI lookups. LogTape doesn't have any such mechanism:

1. [*Structured logging*](https://logtape.org/manual/struct) in LogTape simply organizes data in a structured format rather than plain text. This is a widely adopted industry practice that makes logs more searchable and analyzable.

2. [*Lazy evaluation*](https://logtape.org/manual/struct#lazy-evaluation-of-structured-data) in LogTape doesn't execute arbitrary code or user input. It only calls developer-defined functions when a particular log level is enabled—it's purely a performance optimization to avoid unnecessary calculations for disabled log levels.

There's no mechanism in LogTape that would allow arbitrary code execution through log messages as was the case with Log4Shell.

### On project maturity

> The maintainer could disappear tomorrow, and you will end up with hundreds of call sites to a library that is no longer maintained. 

While LogTape is indeed relatively new (released in April 2024), it has been designed with careful consideration. The codebase is intentionally small with zero dependencies, making it easier to audit and maintain. I'm committed to maintaining the project, and its modular design enables others to contribute or fork if needed.

---

Thank you for the discussion, and I'm happy to address any other questions or concerns about LogTape.

**Comment by iuioiua** [comment]
Deno's gone all in on OpenTelemetry. I suggest we deprecate `@std/log` in favour of OpenTelemetry and any other high-quality third party logging frameworks like LogTape, then remove it. Reminder that this doesn't break existing codebases that use `@std/log`.

**Comment by iuioiua** [comment]
@kt3k WDYT?

**Comment by cristian-spiescu** [comment]
I have a project running an older deno. I planned to upgrade it. But if I'll do, std/log won't be there any more.
Is it OK if I have a new deno, but I still reference an older std/deno?

**Comment by xseman** [comment]
Is there an idea to release @std/log independently from @std? Would be sad to see it gone without continuity..

**Comment by kt3k** [comment]
@cristian-spiescu 
> I have a project running an older deno. I planned to upgrade it. But if I'll do, std/log won't be there any more.
Is it OK if I have a new deno, but I still reference an older std/deno?

While we removed it from github repo, it continues existing in JSR and deno.land. You can keep using either of them.

@xseman 
> Is there an idea to release @std/log independently from @std? Would be sad to see it gone without continuity..

We don’t have a plan to publish it anywhere else. If you feel strongly about it, I’d recommend forking it as a third-party library.
